nwastdlib/asyncio_cache.py

```diff
@@ -0,0 +1,134 @@
+#  Copyright 2019 SURF.
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#          http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+import hashlib
+import hmac
+import pickle  # noqa S403
+import sys
+from functools import wraps
+from typing import Any, Callable
+
+import structlog
+from redis.asyncio import Redis as AIORedis
+
+logger = structlog.get_logger(__name__)
+
+
+def deserialize(data: Any) -> Any:
+    return pickle.loads(data)  # noqa S403
```

Comment:
Not sure if pickle is the correct library to use, It might be worthwhile implementing pickle as a fallback just in case a better serialiser is not available. Furthermore, it might even be better to enable the user to add their own serialiser/serialising functions to make it even more configurable for each use case.

Reply:
I think that pickle (with a HMAC checksum) is a good generic way to implement a LRU cache for generic functions.

A second, less generic, option could be to also implement a JSON backed version. But not every function result can be efficiently serialized to JSON.

The JSON version could also benefit from HMAC signing.

I will try to write a small benchmark, and use that with the current subscriptions cache as an example: to find out know which one is faster.

Reply:
Implemented a nice/simple way to provide your own serializer.